include/session/config/community.hpp

```diff
@@ -0,0 +1,239 @@
+    // one community pubkey (and storing) will affect all communities using the same community
+    // base URL.
+    void set_pubkey(ustring_view pubkey);
+    void set_pubkey(std::string_view pubkey);
+
+    // Accesses the base url (i.e. not including room or pubkey). Always lower-case/normalized.
+    const std::string& base_url() const { return base_url_; }
+
+    // Accesses the room token; this is case-preserving, where possible.  In some contexts, however,
+    // such as volatile info, the case is not preserved and this will always return the normalized
+    // (lower-case) form rather than the preferred form.
+    const std::string& room() const { return localized_room_ ? *localized_room_ : room_; }
+
+    // Accesses the normalized room token, i.e. always lower-case.
+    const std::string& room_norm() const { return room_; }
+
+    const ustring& pubkey() const { return pubkey_; }  // Accesses the server pubkey (32 bytes).
+    std::string pubkey_hex() const;   // Accesses the server pubkey as hex (64 hex digits).
+    std::string pubkey_b32z() const;  // Accesses the server pubkey as base32z (52 alphanumeric
+                                      // digits)
+    std::string pubkey_b64() const;   // Accesses the server pubkey as unpadded base64 (43 from
+                                      // alphanumeric, '+', and '/').
+
+    // Takes a base URL as input and returns it in canonical form.  This involves doing things
+    // like lower casing it and removing redundant ports (e.g. :80 when using http://).  Throws
+    // std::invalid_argument if given an invalid base URL.
+    static std::string canonical_url(std::string_view url);
+
+    // Takes a room token and returns it in canonical form (i.e. lower-cased).  Throws
+    // std::invalid_argument if given an invalid room token (e.g. too long, or containing token
+    // other than a-z, 0-9, -, _).
+    static std::string canonical_room(std::string_view room);
+
+    // Same as above, but modifies the argument in-place instead of returning a modified
+    // copy.
+    static void canonicalize_url(std::string& url);
+    static void canonicalize_room(std::string& room);
+
+    // Takes a full room URL, splits it up into canonical url (see above), room, and server
+    // pubkey.  We take both the deprecated form (e.g.
+    // https://example.com/SomeRoom?public_key=...) and new form
+    // (https://example.com/r/SomeRoom?public_key=...).  The public_key is typically specified
+    // in hex (64 digits), but we also accept base64 (43 chars or 44 with padding) and base32z
+    // (52 chars) encodings (for slightly shorter URLs).
+    //
+    // The returned URL is normalized (lower-cased, and cleaned up).
+    //
+    // The returned room name is *not* normalized, that is, it preserve case.
+    //
+    // Throw std::invalid_argument if anything in the URL is unparseable or invalid.
+    static std::tuple<std::string, std::string, ustring> parse_full_url(std::string_view full_url);
```

Comment:
Can you please add a C interface for parsing a community URL? (figure it would be better to replace the custom Swift function with the libSession one)

Reply:
Added (in `session/config/community.h`)

Reply:
It also has one now to produce a URL from the parts.